man/ftgamma.1
.TH FTGAMMA 1 "February 2023" "FreeType 2.13.0"
.
.
.SH NAME
.
ftgamma \- screen gamma calibration helper
.
.
.SH SYNOPSIS
.
.B ftgamma
.
.
.SH DESCRIPTION
.
The
.B ftgamma
demo program opens a window showing series of gray stripes filled with
solid color and checkered pattern of pixels.  Your monitor's gamma value
roughly corresponds to the horizontal position where the solid and checkered
stripes are equally bright when seen from some distance.  Three brightness
targets are explored: 33%, 50%, and 67%.  If the input-output curve of your
monitor does not follow ideal power law relationship, you might observe
slightly different gamma values at different brightness levels.  It is also
possible to examine gamma values for basic colors.
.
.PP
Two alternative patterns relevant to font rendering are also provided,
where slightly slanted lines are rendered using grayscale and subpixel
anti-aliasing.  In the grayscale case, the correct gamma helps to achieve
uniform perceived thickness of the lines.  The gamma value that corresponds
to vanishing moiré pattern is ideal for anti-aliased font rendering.
In the subpixel case, the correct gamma removes color fringes that remain
even after LCD filtering.
.
.PP
This program does not have any options.
.
.PP
This program is part of the FreeType demos package.
.
.\" eof


ChangeLog
2021-07-18  Werner Lemberg  <wl@gnu.org>

	* Makefile (dist): Ignore `subprojects/freetype2`.

2021-07-18  Werner Lemberg  <wl@gnu.org>

	[ftsdf] Add man page, other minor improvement.

	* man/ftsdf.1: New file.
	* src/fdsdf.c (usage): Resolution can't be changed currently.

2021-07-17  Werner Lemberg  <wl@gnu.org>

	Remove code related to warping.

	This synchronizes the demo programs with commit 2f11522a from the
	FreeType repository.

	* src/ftcommon.c (FTDemo_Hinting_Engine_Change, FTDemo_Get_Info):
	Updated.

	* src/ftdiff.c (ColumnStateRec, render_state_init,
	render_state_draw, event_help, process_event): Updated.

	* src/ftgrid.c (event_help): Updated.

	* src/ftinspect/engine/engine.cpp (Engine::Engine, Engine::update),
	src/ftinspect/engine/engine.hpp (Engine): Updated.
	* src/ftinspect/maingui.cpp (MainGUI::checkHinting,
	MainGUI::checkAutoHinting, MainGUI::createLayout),
	src/ftinspect/maingui.hpp (MainGUI): Updated.

2021-07-01  Werner Lemberg  <wl@gnu.org>

	* src/ftlint.c (main): Minor output improvement.

2021-07-01  Werner Lemberg  <wl@gnu.org>

	* meson.build: Fix version string.

2021-07-01  Werner Lemberg  <wl@gnu.org>

	* subprojects/freetype2.wrap (revision): s/head/HEAD/.

2021-06-30  Werner Lemberg  <wl@gnu.org>

	[ftlint] Improve output.

	* src/ftlint.c (main): Add header to output and remove repetitive
	strings.
	(Analyze): Updated.

2021-06-22  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Improve error reporting.

	* src/ftlint.c (main): Decorate error messages.
	(Error): Take a prefix message argumemt.

2021-06-20  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Use `ft_basename'.

2021-06-20  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Improve error handling.

	* src/ftlint.c (main): Never panic, call `Error', and proceed.
	(Error): Desipher and print an error message instead of...
	(Panic): ...removed.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Fix signed/unsigned mismatch.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Fortify custom indexing.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main, Analyze): Fix signed/unsigned mismatch.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Handle font collections.

	* src/ftlint.c (main): Increment face_index when possible.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Introduce quiet mode, remove extension guessing.

	With so many supported font formats, it does not make sense to guess
	the `ttf' or `ttc' extensions.

	* src/ftlint.c (main): Add `-q' option and improve output formatting;
	remove extension guessing.
	(Usage): Document it.

2021-06-15  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftdump.c (Print_Glyfs, Print_Programs): Handle empty glyphs.

2021-06-14  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Report integral X- and Y-acutance.

	Acutance is a measure of how quickly the intensity changes between
	the minimum and maximum value at the edges.  We calculate is as a
	ratio of the integrals for the absolute second derivative and the
	absolute first derivative.  It equals to 2.0 for bitmap fonts and
	approaches this value for well-hinted fonts.

	* src/ftlint.c (Analyze): Implement it.
	(main): Use it.

2021-06-14  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Report bitmap size, improve output formatting.
	(Checksum): Updated.

2021-06-14  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Implement limited indexing.
	(Usage): Document it.

2021-06-13  Sarthak Bhardwaj  <1sarthakbhardwaj@gmail.com>
            Anurag Thakur  <devanuragthakur@gmail.com>

	[ftlint] Render and calculate MD5 checksum.

	* src/ftlint.c (main): Handle the load flags and the render mode from
	the command line.
	(Checksum): Calculate MD5 sum.
	* src/md5.c, src/md5.h: MD5 source code.
	* Makefile, messon.build: Update the build systems.

2021-06-06  Anuj Verma  <anujv@iitbhilai.ac.in>

	[ftsdf] Synchronize with the change made in the `sdf` renderer.

	* src/ftsdf.c (draw): Change logic to utilize 8bit SDF instead of 16.

2021-06-04  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* graph/gblender.c (gblender_dump_stats): Fix the lookup rate.

2021-06-02  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* graph/grdevice.c (grDoneSurface) [GBLENDER_STATS]: Report stats.
	* graph/gblender.c (gblender_dump_stats): Improve formatting.

2021-05-29  Werner Lemberg  <wl@gnu.org>

	Fix compilation errors and (some) warnings for clang++.

	* graph/gblblit.c (GRGB5[56]5_TO_{RED,GREEN,BLUE}): Add casts.
	(GDST_STOREP): Use `while {} (0)`.
	* graph/gblender.h (GBLENDER_LOOKUP, GBLENDER_LOOKUP_[GBR]): Use
	`while {} (0)`.
	(GBLENDER_CLOSE, GBLENDER_CHANNEL_CLOSE): Remove final semicolon.
	* graph/grblit.c (compose_pixel_full): Use `while {} (0)`.
	(extract5[56]5): Remove final semicolon.
	(blit_gray8_to_24): Fix typo.
	* graph/grfont.c (font): Make it static.

	* src/ftcommon.c (FTDemo_Icon): Don't use obfuscating comma
	operator.
	* src/ftsdf.c (main): Move definition of `flip_y` before `goto`
	instructions.

2021-05-28  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[graph/win32] Fix double-dealing of some keystrokes.

	* graph/grwin32.c (gr_win32_surface_listen_event): Deprioritize
	`TranslateMessage' to avoid double-dealing of Tab, Esc, Return.

2021-05-23  David Turner  <david@freetype.org>

	* src/ftmemchk.c: Remove minor compiler warning.

	Casting a pointer to (long) isn't always correct.

2021-05-23  David Turner  <david@freetype.org>

	* src/ftpngout.c: Remove compiler warning for Mingw64 toolchain.

	The #pragma statement is only valid with Microsoft Visual C++

2021-05-23  David Turner  <david@freetype.org>

	* graph/win32/grwin32.c: Fix compiler warnings.

	- Redefine the LCS_sRGB macro to properly compile with the Mingw64
	 toolchain and the -Wmultichar warning enabled.


src/ftgamma.c
/****************************************************************************/
/*                                                                          */
/*  The FreeType project -- a free and portable quality TrueType renderer.  */
/*                                                                          */
/*  Copyright (C) 2004-2023 by                                              */
/*  D. Turner, R.Wilhelm, and W. Lemberg                                    */
/*                                                                          */
/*                                                                          */
/*  ftgamma - gamma matcher                                                 */
/*                                                                          */
/****************************************************************************/


#include "ftcommon.h"
#include <math.h>


  static FTDemo_Display*  display;

  static grBitmap   bit1 = { 300, 600, 600, gr_pixel_mode_gray, 256, NULL };
  static grBitmap   bit2 = { 288, 600, 600, gr_pixel_mode_gray, 256, NULL };
  static int        status = 0;


  static void
  do_ptrn( grBitmap*  bitmap,
           int        x,
           int        y,
           int        w,
           int        h )
  {
    int     pitch = bitmap->pitch;
    int     i, j, k;
    double  p[4];

    unsigned char*  line;

    for ( i = 0; i < h; i++ )
    {
      for ( k = 0; k < 4; k++)
      {
        j = 2 * i + 1 + ( k - 4 ) * h / 2;
        if ( j > h )
          j -= 2 * h;
        if ( j < -h )
          j += 2 * h;
        if ( j < 0 )
          j = -j;
        j -= h / 4;
        if ( j < 0 )
          j = 0;
        if ( j > h / 2 )
          j = h / 2;

        p[k] = 2. * j / h;
      }

      line = bitmap->buffer + ( y + i ) * pitch + x;
      for ( j = 0, k = 0; j < w; j++ )
      {
        line[j] = (unsigned char)( 0.5 +
                                   255. * pow ( p[k],
                                                1. / (1. + 2. * j / w ) ) );
        k++;
        if ( k == 4 )
          k = 0;
      }
    }
  }


  static FT_Error
  GammaPtrn( grBitmap*  bitmap )
  {
    int  x = 0;
    int  y = 0;
    int  h = ( bitmap->rows - 2 * y ) / 2;
    int  w = bitmap->width - 2 * x;


    do_ptrn( bitmap, x,    y, w, h );
    do_ptrn( bitmap, x, y+=h, w, h );

    return 0;
  }


  static void
  do_fill( grBitmap*  bitmap,
           int        x,
           int        y,
           int        w,
           int        h,
           int        back,
           int        fore )
  {
    int     pitch = bitmap->pitch;
    int     i;
    double  b, f;

    unsigned char*  line = bitmap->buffer + y*pitch + x;


    if ( back == 0 || back == 255 )
      for ( i = 0; i < w; i++ )
        line[i + ( i & 1 ) * pitch] = (unsigned char)back;
    else
      for ( b = back / 255., i = 0; i < w; i++ )
        line[i + ( i & 1 ) * pitch] =
          (unsigned char)( 0.5 + 255. * pow ( b, 1. / (1. + 2. * i / w ) ) );

    if ( fore == 0 || fore == 255 )
      for ( i = 0; i < w; i++ )
        line[i + ( ~i & 1 ) * pitch] = (unsigned char)fore;
    else
      for ( f = fore / 255., i = 0; i < w; i++ )
        line[i + ( ~i & 1 ) * pitch] =
          (unsigned char)( 0.5 + 255. * pow ( f, 1. / (1. + 2. * i / w ) ) );

    for ( i = 2; i < h; i += 2 )
    {
      memcpy( line + i * pitch, line, (size_t)w );
      memcpy( line + i * pitch + pitch, line + pitch, (size_t)w );
    }
  }


  static FT_Error
  GammaGrid( grBitmap*  bitmap )
  {
    int  x = 0;
    int  y = 0;
    int  h = ( bitmap->rows - 2 * y ) / 15;
    int  w = bitmap->width - 2 * x;


    do_fill( bitmap, x,    y, w, h,  85, 255 );
    do_fill( bitmap, x, y+=h, w, h, 170, 170 );
    do_fill( bitmap, x, y+=h, w, h,  85, 255 );
    do_fill( bitmap, x, y+=h, w, h, 170, 170 );
    do_fill( bitmap, x, y+=h, w, h,  85, 255 );

    do_fill( bitmap, x, y+=h, w, h,   0, 255 );
    do_fill( bitmap, x, y+=h, w, h, 127, 127 );
    do_fill( bitmap, x, y+=h, w, h,   0, 255 );
    do_fill( bitmap, x, y+=h, w, h, 127, 127 );
    do_fill( bitmap, x, y+=h, w, h,   0, 255 );

    do_fill( bitmap, x, y+=h, w, h,   0, 170 );
    do_fill( bitmap, x, y+=h, w, h,  85,  85 );
    do_fill( bitmap, x, y+=h, w, h,   0, 170 );
    do_fill( bitmap, x, y+=h, w, h,  85,  85 );
    do_fill( bitmap, x, y+=h, w, h,   0, 170 );

    return 0;
  }


  static void
  event_help( void )
  {
    grEvent  dummy_event;


    FTDemo_Display_Clear( display );
    grSetLineHeight( 10 );
    grGotoxy( 0, 0 );
    grSetMargin( 2, 1 );
    grGotobitmap( display->bitmap );


    grWriteln( "FreeType Gamma Matcher" );
    grLn();
    grWriteln( "Use the following keys:" );
    grLn();
    grWriteln( "F1, ?       display this help screen" );
    grLn();
    grWriteln( "space       cycle through color");
    grWriteln( "tab         alternate patterns");
    grWriteln( "G           show gamma ramp" );
    grLn();
    grLn();
    grWriteln( "press any key to exit this help screen" );

    grRefreshSurface( display->surface );
    grListenSurface( display->surface, gr_event_key, &dummy_event );
  }


  static void
  event_color_change( void )
  {
    static int     i = 7;
    unsigned char  r = i & 4 ? 0xff : 0;
    unsigned char  g = i & 2 ? 0xff : 0;
    unsigned char  b = i & 1 ? 0xff : 0;


    display->back_color = grFindColor( display->bitmap, 0, 0, 0, 0xff );
    display->fore_color = grFindColor( display->bitmap, r, g, b, 0xff );

    i++;
    if ( ( i & 0x7 ) == 0 )
      i = 1;
  }


  static void
  event_gamma_grid( void )
  {
    grEvent  dummy_event;
    int      g;
    int      yside  = 11;
    int      xside  = 10;
    int      levels = 17;
    int      gammas = 30;
    int      x_0    = ( display->bitmap->width - levels * xside ) / 2;
    int      y_0    = ( display->bitmap->rows - gammas * ( yside + 1 ) ) / 2;
    int      pitch  = display->bitmap->pitch;


    FTDemo_Display_Clear( display );
    grGotobitmap( display->bitmap );

    if ( pitch < 0 )
      pitch = -pitch;

    memset( display->bitmap->buffer,
            100,
            (size_t)pitch * (size_t)display->bitmap->rows );

    grWriteCellString( display->bitmap, 0, 0, "Gamma grid",
                       display->fore_color );


    for ( g = 1; g <= gammas; g++ )
    {
      double  ggamma = 0.1 * g;
      char    temp[6];
      int     y = y_0 + ( yside + 1 ) * ( g - 1 );
      int     nx;


      snprintf( temp, sizeof ( temp ), "%.1f", ggamma );
      grWriteCellString( display->bitmap, x_0 - 32, y + ( yside - 6 ) / 2,
                         temp, display->fore_color );

      for ( nx = 0; nx < levels; nx++ )
      {
        double   p  = nx / (double)( levels - 1 );
        int      gm = (int)( 255.0 * pow( p, ggamma ) + 0.5 );
        grColor  c;


        c = grFindColor( display->bitmap, gm, gm, gm, 0xff );
        grFillRect( display->bitmap, x_0 + nx * xside, y, xside, yside, c );
      }
    }

    grRefreshSurface( display->surface );
    grListenSurface( display->surface, gr_event_key, &dummy_event );
  }


  static void
  Render_Bitmap( grBitmap*  out,
                 grBitmap*  in,
                 int        x,
                 int        y,
                 grColor    color,
                 int        lcd )
  {
    int  pitch = abs( out->pitch );
    int  l = 0, r = in->width, t = 0, b = in->rows;
    int  i, ii, j;

    unsigned char*  src;
    unsigned char*  dst;


    /* clip if necessary */
    if ( x < 0 )
      l = -x;
    if ( y < 0 )
      t = -y;
    if ( x + r > out->width )
      r = out->width - x;
    if ( y + b > out->rows )
      b = out->rows - y;

    if ( color.chroma[0] == 255 )
      for ( i = t; i < b; i++ )
      {
        ii = ( i +  0 * lcd ) % in->rows;
        src = in->buffer + ii * in->pitch + l;
        dst = out->buffer + ( y + i ) * pitch + 3 * ( x + l );
        for ( j = l; j < r; j++, src++, dst += 3 )
          *dst = *src;
      }

    if ( color.chroma[1] == 255 )
      for ( i = t; i < b; i++ )
      {
        ii = ( i + 12 * lcd ) % in->rows;
        src = in->buffer + ii * in->pitch + l;
        dst = out->buffer + ( y + i ) * pitch + 3 * ( x + l ) + 1;
        for ( j = l; j < r; j++, src++, dst += 3 )
          *dst = *src;
      }

    if ( color.chroma[2] == 255 )
      for ( i = t; i < b; i++ )
      {
        ii = ( i + 24 * lcd ) % in->rows;
        src = in->buffer + ii * in->pitch + l;
        dst = out->buffer + ( y + i ) * pitch + 3 * ( x + l ) + 2;
        for ( j = l; j < r; j++, src++, dst += 3 )
          *dst = *src;
      }
  }


  static int
  Process_Event( void )
  {
    grEvent  event;
    int      ret = 0;


    grRefreshSurface( display->surface );
    grListenSurface( display->surface, 0, &event );

    if ( event.type == gr_event_resize )
      return ret;

    switch ( event.key )
    {
    case grKeyEsc:
    case grKEY( 'q' ):
      ret = 1;
      break;

    case grKeyF1:
    case grKEY( '?' ):
      event_help();
      break;

    case grKeySpace:
      event_color_change();
      break;

    case grKeyTab:
      status++;
      if ( status > 2 )
        status = 0;
      break;

    case grKEY( 'G' ):
      event_gamma_grid();
      break;

    default:
      break;
    }

    return ret;
  }


  int
  main( void )
  {
    char  buf[4];
    int   i;

    display = FTDemo_Display_New( NULL, DIM "x24",
                        "FreeType Gamma Matcher - press ? for help" );
    if ( !display )
    {
      PanicZ( "could not allocate display surface" );
    }

    grNewBitmap( bit1.mode, bit1.grays, bit1.width, bit1.rows, &bit1 );
    GammaGrid( &bit1 );

    grNewBitmap( bit2.mode, bit2.grays, bit2.width, bit2.rows, &bit2 );
    GammaPtrn( &bit2 );

    event_color_change();

    do
    {
      int  x = display->bitmap->width / 2;
      int  y = display->bitmap->rows / 2;


      FTDemo_Display_Clear( display );

      switch ( status )
      {
      case 0:
        grWriteCellString( display->bitmap, x - 84, y - 165,
                           "Solid-Checkered Pattern", display->fore_color );
        Render_Bitmap( display->bitmap, &bit1, x - 300, y - 150,
                       display->fore_color, 0 );
        break;
      case 1:
        grWriteCellString( display->bitmap, x - 84, y - 165,
                           "Grayscale Anti-Aliasing", display->fore_color );
        Render_Bitmap( display->bitmap, &bit2, x - 300, y - 144,
                       display->fore_color, 0 );
        break;
      case 2:
        grWriteCellString( display->bitmap, x - 84, y - 165,
                           "Subpixel  Anti-Aliasing", display->fore_color );
        Render_Bitmap( display->bitmap, &bit2, x - 300, y - 144,
                       display->fore_color, 1 );
        break;
      }

      for ( i = 0; i <= 10; i++ )
      {
        snprintf( buf, sizeof ( buf ), "%.1f", 1. + .2 * i );
        grWriteCellString( display->bitmap, x - 311 + i * 60, y + 155,
                           buf, display->fore_color );
      }

      grWriteCellString( display->bitmap, x - 20, y + 170,
                         "Gamma", display->fore_color );
    } while ( Process_Event() == 0 );

    FTDemo_Display_Done( display );

    exit( 0 );      /* for safety reasons */
    /* return 0; */ /* never reached */
  }


/* End */
